### PR TITLE
UIP-2414: Resolve ddc test failures

### PR DESCRIPTION
**Pull request by clairesarsam-wf**
# Ultimate Problem

Several dozen tests failed when the test code was compiled using `dartdevc`.

# Solution

* Resolve failures
* Only a few changes ended up needing to be made.
  * [Patch the name property on emulated functions bug to fix the behavior of the `name`/`type` fields](https://github.com/cleandart/react-dart/pull/122/files#diff-12ac054bb8b76f35c1035afef1d94396R361)
  * [Two tests cannot be run on ddc currently due to the `noSuchMethod` on emulated functions bug](https://github.com/cleandart/react-dart/pull/122/files#diff-7a382a75d4df367a3f5df34d837d9227R1)
  * [Resolve runtime type casting errors](https://github.com/cleandart/react-dart/pull/122/files#diff-c00a02aa016b25a86fb30a2093f935a1L430)

# How to +10/QA

1. All tests should pass when run normally:
```sh
pub run test -p content-shell,chrome
```
Note: I had to do this on `1.23.0`.
2. All tests should pass when run like this:
First add the following to pubspec.yaml:
```yaml
transformers:
- test/pub_serve:
    $include: test/**_test.dart
```
Then, run:
```sh
$ pub serve test --web-compiler=dartdevc
# in a second window
pub run test -p chrome --pub-serve=8080 -x ddcFailure
```
---

Please review @aaronlademann-wf @greglittlefield-wf @jacehensley-wf

### CODE REVIEW

File: test/factory/common_factory_tests.dart [hunk 1]
```diff
@@ -33,7 +35,7 @@ void commonFactoryTests(Function factory) {
     test('multiple arguments', () {
       var instance = factory({}, 'one', 'two');
       expect(getJsChildren(instance), equals(['one', 'two']));
-    });
+    }, tags: 'ddcFailure');
```
Comment (greglittlefield-wf) [inline test/factory/common_factory_tests.dart]: This should have a comment as to why it's disabled in the DDC

File: test/factory/common_factory_tests.dart [hunk 2]
```diff
@@ -155,7 +157,7 @@ void _childKeyWarningTests(Function factory) {
       );
 
       expect(consoleErrorCalled, isFalse, reason: 'should not have outputted a warning');
-    });
+    }, tags: 'ddcFailure');
```
Comment (greglittlefield-wf) [inline test/factory/common_factory_tests.dart]: This should have a comment as to why it's disabled in the DDC

File: test/factory/common_factory_tests.dart [hunk 3]
```diff
@@ -1,3 +1,5 @@
+@Tags(const ["ddcFailure"])
```
Comment (greglittlefield-wf) [inline test/factory/common_factory_tests.dart]: Is this needed? Wouldn't this mark the whole suite as `ddcFailure`?

Comment (clairesarsam-wf) [inline test/factory/common_factory_tests.dart]: I think I misread the docs about this.. for some reason I thought I needed to declare the tag to be able to use it. I'll take this out.

**Comment by clairesarsam-wf** [comment]
@greglittlefield-wf Feedback addressed

**Comment by jacehensley-wf** [comment]
+1

**Comment by greglittlefield-wf** [comment]
- [x] All tests pass when run normally in content_shell and dart2js in Dart 1.23.0
- [x] All tests pass when run in the DDC, as per instructions
- [x] Smoke tested rendering in WSD examples, no regressions found

QA +10

**Pull request merged**